File: monytix/lib/models/goals_models.dart
/// Data models for Goals Questionnaire

class LifeContext {
  final String ageBand;
  final bool dependentsSpouse;
  final int dependentsChildrenCount;
  final bool dependentsParentsCare;
  final String housing;
  final String employment;
  final String incomeRegularity;
  final String regionCode;
  final bool emergencyOptOut;

  LifeContext({
    required this.ageBand,
    this.dependentsSpouse = false,
    this.dependentsChildrenCount = 0,
    this.dependentsParentsCare = false,
    required this.housing,
    required this.employment,
    required this.incomeRegularity,
    required this.regionCode,
    this.emergencyOptOut = false,
  });

  factory LifeContext.fromJson(Map<String, dynamic> json) {
    return LifeContext(
      ageBand: json['age_band'] ?? '',
      dependentsSpouse: json['dependents_spouse'] ?? false,
      dependentsChildrenCount: json['dependents_children_count'] ?? 0,
      dependentsParentsCare: json['dependents_parents_care'] ?? false,
      housing: json['housing'] ?? '',
      employment: json['employment'] ?? '',
      incomeRegularity: json['income_regularity'] ?? '',
      regionCode: json['region_code'] ?? '',
      emergencyOptOut: json['emergency_opt_out'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'age_band': ageBand,
      'dependents_spouse': dependentsSpouse,
      'dependents_children_count': dependentsChildrenCount,
      'dependents_parents_care': dependentsParentsCare,
      'housing': housing,
      'employment': employment,
      'income_regularity': incomeRegularity,
      'region_code': regionCode,
      'emergency_opt_out': emergencyOptOut,
    };
  }
}

class GoalCatalogItem {
  final String goalCategory;
  final String goalName;
  final String defaultHorizon;
  final String policyLinkedTxnType;
  final bool isMandatoryFlag;
  final String? suggestedMinAmountFormula;
  final int displayOrder;

  GoalCatalogItem({
    required this.goalCategory,
    required this.goalName,
    required this.defaultHorizon,
    required this.policyLinkedTxnType,
    required this.isMandatoryFlag,
    this.suggestedMinAmountFormula,
    required this.displayOrder,
  });

  factory GoalCatalogItem.fromJson(Map<String, dynamic> json) {
    return GoalCatalogItem(
      goalCategory: json['goal_category'] ?? '',
      goalName: json['goal_name'] ?? '',
      defaultHorizon: json['default_horizon'] ?? '',
      policyLinkedTxnType: json['policy_linked_txn_type'] ?? '',
      isMandatoryFlag: json['is_mandatory_flag'] ?? false,
      suggestedMinAmountFormula: json['suggested_min_amount_formula'],
      displayOrder: json['display_order'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'goal_category': goalCategory,
      'goal_name': goalName,
      'default_horizon': defaultHorizon,
      'policy_linked_txn_type': policyLinkedTxnType,
      'is_mandatory_flag': isMandatoryFlag,
      'suggested_min_amount_formula': suggestedMinAmountFormula,
      'display_order': displayOrder,
    };
  }

  String get key => '$goalCategory:$goalName';
}

class SelectedGoal {
  final String goalCategory;
  final String goalName;
  final double estimatedCost;
  final DateTime? targetDate;
  final double currentSavings;
  final int importance;
  final String? notes;

  SelectedGoal({
    required this.goalCategory,
    required this.goalName,
    required this.estimatedCost,
    this.targetDate,
    this.currentSavings = 0.0,
    this.importance = 3,
    this.notes,
  });

  SelectedGoal copyWith({
    String? goalCategory,
    String? goalName,
    double? estimatedCost,
    DateTime? targetDate,
    double? currentSavings,
    int? importance,
    String? notes,
  }) {
    return SelectedGoal(
      goalCategory: goalCategory ?? this.goalCategory,
      goalName: goalName ?? this.goalName,
      estimatedCost: estimatedCost ?? this.estimatedCost,
      targetDate: targetDate ?? this.targetDate,
      currentSavings: currentSavings ?? this.currentSavings,
      importance: importance ?? this.importance,
      notes: notes ?? this.notes,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'goal_category': goalCategory,
      'goal_name': goalName,
      'estimated_cost': estimatedCost,
      'target_date': targetDate?.toIso8601String().split('T')[0],
      'current_savings': currentSavings,
      'importance': importance,
      'notes': notes,
    };
  }
}

File: apk_monytix/app/src/main/java/com/example/apk/goals/GoalsModels.kt
package com.example.apk.goals

import com.google.gson.annotations.SerializedName
import java.util.UUID

// MARK: - Goal Catalog

data class GoalCatalogItem(
    @SerializedName("goal_category") val goalCategory: String,
    @SerializedName("goal_name") val goalName: String,
    @SerializedName("default_horizon") val defaultHorizon: String,
    @SerializedName("policy_linked_txn_type") val policyLinkedTxnType: String,
    @SerializedName("is_mandatory_flag") val isMandatoryFlag: Boolean,
    @SerializedName("suggested_min_amount_formula") val suggestedMinAmountFormula: String?,
    @SerializedName("display_order") val displayOrder: Int
)

// MARK: - Life Context

data class LifeContext(
    @SerializedName("age_band") val ageBand: String,
    @SerializedName("dependents_spouse") val dependentsSpouse: Boolean,
    @SerializedName("dependents_children_count") val dependentsChildrenCount: Int,
    @SerializedName("dependents_parents_care") val dependentsParentsCare: Boolean,
    @SerializedName("housing") val housing: String,
    @SerializedName("employment") val employment: String,
    @SerializedName("income_regularity") val incomeRegularity: String,
    @SerializedName("region_code") val regionCode: String,
    @SerializedName("emergency_opt_out") val emergencyOptOut: Boolean,
    @SerializedName("monthly_investible_capacity") val monthlyInvestibleCapacity: Double? = null,
    @SerializedName("total_monthly_emi_obligations") val totalMonthlyEMIObligations: Double? = null,
    @SerializedName("risk_profile_overall") val riskProfileOverall: String? = null,
    @SerializedName("review_frequency") val reviewFrequency: String? = null,
    @SerializedName("notify_on_drift") val notifyOnDrift: Boolean? = null,
    @SerializedName("auto_adjust_on_income_change") val autoAdjustOnIncomeChange: Boolean? = null
)

// MARK: - Selected Goal

data class SelectedGoal(
    @SerializedName("goal_category") val goalCategory: String,
    @SerializedName("goal_name") val goalName: String,
    @SerializedName("estimated_cost") var estimatedCost: Double = 0.0,
    @SerializedName("target_date") var targetDate: String? = null,
    @SerializedName("current_savings") var currentSavings: Double = 0.0,
    @SerializedName("importance") var importance: Int = 3,
    @SerializedName("notes") var notes: String? = null
)

// MARK: - Goal Progress

data class GoalProgress(
    @SerializedName("goal_id") val goalId: String,
    @SerializedName("goal_name") val goalName: String,
    @SerializedName("progress_pct") val progressPct: Double,
    @SerializedName("current_savings_close") val currentSavingsClose: Double,
    @SerializedName("remaining_amount") val remainingAmount: Double,
    @SerializedName("projected_completion_date") val projectedCompletionDate: String?,
    @SerializedName("milestones") val milestones: List<Int>
)

data class GoalsProgressResponse(
    @SerializedName("goals") val goals: List<GoalProgress>
)

// MARK: - Submit Request/Response

data class GoalsSubmitRequest(
    @SerializedName("context") val context: LifeContext,
    @SerializedName("selected_goals") val selectedGoals: List<SelectedGoal>
)

data class GoalsSubmitResponse(
    @SerializedName("goals_created") val goalsCreated: List<GoalCreatedItem>
)

data class GoalCreatedItem(
    @SerializedName("goal_id") val goalId: String,
    @SerializedName("priority_rank") val priorityRank: Int?
)

// MARK: - Catalog Response

data class GoalCatalogResponse(
    @SerializedName("goals") val goals: List<GoalCatalogItem>
)

// MARK: - Life Context Response

data class LifeContextResponse(
    @SerializedName("context") val context: LifeContext?
)
